### PR TITLE
Adding relative import support behind a settings toggle

### PR DESCRIPTION
**Pull request by davidsiems**
# Relative Import Paths
Adding support for relative paths in import declarations for ns1 and ns2 scripts. Functionality is gated behind a new setting in the settings menu and defaults to off.

Screen Shot 2022-01-23 at 12 21 44 AM

# Testing
Verified that behavior before the change matches behavior after the change with the setting disabled.
Verified that with the setting enabled relative import paths work from the root and sub-directories for both ns1 and ns2 scripts.
Verified static RAM calculations are accurate in all cases.

### Cases tested
[path-tests.zip](https://github.com/danielyxie/bitburner/files/7920061/path-tests.zip)

### Results
```
From Root Directory (NS1)
       Path      |  Before | After-Off | After-On
root-import           Y          Y          Y
/root-import          N          N          Y
./root-import         Y          Y          Y
../root-import        N          N          Y
sub/sub-import        N          N          Y
/sub/sub-import       Y          Y          Y
./sub/sub-import      N          N          Y
../sub/sub-import     N          N          Y
sub-import            N          N          N
/sub-import           N          N          N
./sub-import          N          N          N
../sub-import         N          N          N


From Sub Directory (NS1)
       Path      |  Before | After-Off | After-On
root-import           Y          Y          N
/root-import          N          N          Y
./root-import         Y          Y          N
../root-import        N          N          Y
sub/sub-import        N          N          N
/sub/sub-import       Y          Y          Y
./sub/sub-import      N          N          N
../sub/sub-import     N          N          Y
sub-import            N          N          Y
/sub-import           N          N          N
./sub-import          N          N          Y
../sub-import         N          N          N 



From Root Directory (NS2)
       Path      |  Before | After-Off | After-On
root-import           Y          Y          Y
/root-import          Y          Y          Y
./root-import         Y          Y          Y
../root-import        N          N          Y
sub/sub-import        Y          Y          Y
/sub/sub-import       Y          Y          Y
./sub/sub-import      Y          Y          Y
../sub/sub-import     N          N          Y
sub-import            N          N          N
/sub-import           N          N          N
./sub-import          N          N          N
../sub-import         N          N          N


From Sub Directory (NS2)
       Path      |  Before | After-Off | After-On
root-import           Y          Y          N
/root-import          Y          Y          Y
./root-import         Y          Y          N
../root-import        N          N          Y
sub/sub-import        Y          Y          N
/sub/sub-import       Y          Y          Y
./sub/sub-import      Y          Y          N
../sub/sub-import     N          N          Y
sub-import            N          N          Y
/sub-import           N          N          N
./sub-import          N          N          Y
../sub-import         N          N          N 
```

### CODE REVIEW

File: src/Terminal/DirectoryHelpers.ts [hunk 1]
```diff
@@ -316,3 +319,20 @@ export function areImportsEquals(f0: string, f1: string): boolean {
   if (!f1.endsWith(".ns") && !f1.endsWith(".js")) f1 = f1 + ".js";
   return areFilesEqual(f0, f1);
 }
+
+export function resolveImportPath(parentScriptPath: string, importedScriptPath: string): string {
```
Comment (theit8514) [inline src/Terminal/DirectoryHelpers.ts:323]: This could use an doc comment describing what the variables are. What does path.dirname do if the folder path is given rather than the full path to the script?

File: src/Script/RamCalculations.ts [hunk 2]
```diff
@@ -13,11 +13,11 @@ import { RamCalculationErrorCode } from "./RamCalculationErrorCodes";
 import { RamCosts, RamCostConstants } from "../Netscript/RamCostGenerator";
 import { Script } from "../Script/Script";
 import { WorkerScript } from "../Netscript/WorkerScript";
-import { areImportsEquals } from "../Terminal/DirectoryHelpers";
+import { areImportsEquals, resolveImportPath } from "../Terminal/DirectoryHelpers";
 import { IPlayer } from "../PersonObjects/IPlayer";
 
 export interface RamUsageEntry {
-  type: 'ns' | 'dom' | 'fn' | 'misc';
+  type: "ns" | "dom" | "fn" | "misc";
```
Comment (theit8514) [inline src/Script/RamCalculations.ts:20]: Unrelated formatting changes can probably be excluded.

File: src/Script/RamCalculations.ts [hunk 3]
```diff
@@ -49,6 +49,7 @@ async function parseOnlyRamCalculate(
   otherScripts: Script[],
   code: string,
   workerScript: WorkerScript,
+  scriptPath: string,
```
Comment (theit8514) [inline src/Script/RamCalculations.ts:52]: Parameter needs to be added to the doc comment. Should indicate that it is the full path to the script that is currently in the code parameter.

File: src/Script/RamCalculations.ts [hunk 4]
```diff
@@ -410,6 +412,7 @@ export async function calculateRamUsage(
   player: IPlayer,
   codeCopy: string,
   otherScripts: Script[],
+  scriptPath: string,
```
Comment (theit8514) [inline src/Script/RamCalculations.ts:415]: Add doc comment

**Review by theit8514 — COMMENTED** [review]
Looks good besides some formatting changes that make it harder to review.

One downside to the checkbox will be that people will find a set of scripts that they would expect to work and other people claim work fine for them, but they are on an opposite import setting.

**Comment by davidsiems** [comment]
> Looks good besides some formatting changes that make it harder to review.
> 
> One downside to the checkbox will be that people will find a set of scripts that they would expect to work and other people claim work fine for them, but they are on an opposite import setting.

Yeah - Tried to split those out into a separate commit at least. Could fully remove. FWIW Running npm run format is currently hitting roughly 100 files.
